File: NQueens.java
import java.util.*;

public class NQueens {

    static int N;

    // Check if it's safe to place queen at (row, col)
    static boolean isSafe(int[][] board, int row, int col) {

        // Check column
        for (int i = 0; i < row; i++) {
            if (board[i][col] == 1)
                return false;
        }

        // Check left diagonal
        for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--) {
            if (board[i][j] == 1)
                return false;
        }

        // Check right diagonal
        for (int i = row - 1, j = col + 1; i >= 0 && j < N; i--, j++) {
            if (board[i][j] == 1)
                return false;
        }

        return true;
    }

    // Backtracking function
    static boolean solve(int[][] board, int row) {

        // Base case: all queens placed
        if (row == N) {
            return true;
        }

        // Try all columns in current row
        for (int col = 0; col < N; col++) {

            if (isSafe(board, row, col)) {

                board[row][col] = 1; // place queen

                if (solve(board, row + 1)) {
                    return true;
                }

                board[row][col] = 0; // backtrack
            }
        }

        return false;
    }

    // Print solution
    static void printBoard(int[][] board) {
        System.out.println("Solution:");
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                System.out.print(board[i][j] + " ");
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter value of N: ");
        N = sc.nextInt();

        int[][] board = new int[N][N];

        if (solve(board, 0)) {
            printBoard(board);
        } else {
            System.out.println("No solution exists");
        }

        sc.close();
    }
}

File: nqueens.py
MAX = 20

# Initialize variables
N = int(input("Enter number of queens (N): "))

board = [[0]*N for _ in range(N)]

column = [False]*N
leftDiagonal = [False]*(2*N)
rightDiagonal = [False]*(2*N)

# Print solution
def print_solution():
    print("\nSolution:")
    for i in range(N):
        for j in range(N):
            print(board[i][j], end=" ")
        print()

# Solve using Backtracking + Branch & Bound
def solve_n_queens(row):
    if row >= N:
        print_solution()
        return True

    for col in range(N):

        if (not column[col] and
            not leftDiagonal[row - col + N - 1] and
            not rightDiagonal[row + col]):

            # Place queen
            board[row][col] = 1
            column[col] = True
            leftDiagonal[row - col + N - 1] = True
            rightDiagonal[row + col] = True

            # Recur
            if solve_n_queens(row + 1):
                return True

            # Backtrack
            board[row][col] = 0
            column[col] = False
            leftDiagonal[row - col + N - 1] = False
            rightDiagonal[row + col] = False

    return False

# Run
if not solve_n_queens(0):
    print("No solution exists.")